Speed up model parallel initialization

alexqdh opened the pull request:
The process groups are created on all ranks regardless of whether the current rank was actually part of that group. This is extremely time-consuming, especially during large-scale distributed training. By default, in torch groups should be created in the same order in all processes. 

Parameter `use_local_synchronization=True` is added to all `create_group()` calls to enable local synchronization optimization, which can further improve performance during group creation. After modification, process groups are now only created when the current rank is actually part of the group `(if rank in ranks:)`.

Phlip79 commented:
We are changing our review process and marking all open, unlabeled PRs as draft. This change will go in effect starting once #3659 is merged.

Moving forward, all PRs will be required to start as draft PRs. If you wish to get your PR merged, mark your PR as “Ready for review”. Read more about the new process at [submit.md](https://github.com/Phlip79/Megatron-LM/blob/d3763629bcdd04acca52477af7eb8795090426ff/docs/developer/submit.md).

alexqdh committed:
perf(parallel_state): speed up process group init with local sync

- Call create_group only from ranks that belong to each group
- Pass use_local_synchronization=True for NCCL/gloo new_group paths
- Apply to hierarchical CP, hybrid DP-CP helpers, and all MPU groups

Made-with: Cursor

alexqdh committed:
Merge branch 'main' into speed_up_mp_initialization

ericharper commented:
Closing since no response. Please re-open if still needed.

alexqdh commented:
@ericharper  Thanks for you reply. This PR is still needed. I rebased the branch on latest main and resolved the conflicts. Could you help reopen it?
